frame/support/procedural/src/expand_after.rs

```diff
@@ -0,0 +1,131 @@
+		},
+	}
+}
+
+struct ExpandAfterDef {
+	// Pattern to expand after, this is ensured to have no TokenTree::Group nor TokenTree::Literal
+	// (i.e. contains only Punct or Ident), and not being empty.
+	expand_after: Vec<TokenTree>,
+	// Token stream to write after match.
+	expand_with: TokenStream,
+	// Token stream to search and write inside.
+	expand_in: TokenStream,
+}
+
+impl syn::parse::Parse for ExpandAfterDef {
+	fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+		let expand_after;
+		let _replace_with_bracket: syn::token::Brace = syn::braced!(expand_after in input);
+		let expand_after =
+			expand_after.parse::<TokenStream>()?.into_iter().collect::<Vec<TokenTree>>();
+
+		if let Some(t) = expand_after.iter().find(|t| matches!(t, TokenTree::Group(_))) {
+			return Err(syn::Error::new(t.span(), "Unexpected group token tree"))
+		}
+		if let Some(t) = expand_after.iter().find(|t| matches!(t, TokenTree::Literal(_))) {
+			return Err(syn::Error::new(t.span(), "Unexpected literal token tree"))
+		}
+
+		if expand_after.is_empty() {
+			return Err(syn::Error::new(Span::call_site(), "empty match pattern is invalid"))
+		}
+
+		let expand_with;
+		let _replace_with_bracket: syn::token::Brace = syn::braced!(expand_with in input);
+		let expand_with: TokenStream = expand_with.parse()?;
+
+		Ok(Self { expand_with, expand_after, expand_in: input.parse()? })
+	}
+}
+
+// Replace the first found `after` pattern by content of `with`.
+// `with` must be some (Option is used for internal simplification).
+// `after` musn't be empty and only contains Ident or Punct
+fn expand_in_stream(
+	after: &[TokenTree],
+	with: &mut Option<TokenStream>,
+	stream: TokenStream,
+) -> Result<TokenStream, ()> {
+	assert!(with.is_some(), "`with` must be some, Option is used because `with` is used only once");
+	assert!(!after.is_empty(), "`after` mustn't be empty, otherwise it cannot be found");
+
+	let mut stream = stream.into_iter();
+	let mut extended = TokenStream::new();
+	let mut match_cursor = 0;
+
+	loop {
+		match stream.next() {
+			Some(TokenTree::Group(group)) => {
+				match_cursor = 0;
+				let stream = group.stream();
+				match expand_in_stream(after, with, stream) {
+					Ok(stream) => {
+						extended
+							.extend(once(TokenTree::Group(Group::new(group.delimiter(), stream))));
+						break
+					},
+					Err(_) => {
+						extended.extend(once(TokenTree::Group(group)));
+					},
+				}
+			},
+			Some(other) => {
+				use TokenTree::{Ident, Punct};
+
+				let other_match_pattern = match (&other, &after[match_cursor]) {
+					(Ident(i1), Ident(i2)) => i1 == i2,
+					(Punct(p1), Punct(p2)) => p1.as_char() == p2.as_char(),
+					_ => false,
+				};
+
+				if other_match_pattern {
+					match_cursor += 1
+				}
+
+				extended.extend(once(other));
+
+				if match_cursor == after.len() {
+					extended.extend(once(with.take().expect("with is used to replace only once")));
+					break
+				}
```

Comment:
I wonder if there's a way to make this more self-documenting? Perhaps we can extract this algorithm to a function and name it `advance_match_cursor` or something...